[과제] Restaurant Details Page

TaeWoongKim2 opened the pull request:
## 학습목표 - 레스토랑 상세 페이지 구현하기

### 과제 목표

- [x] TDD로 기능 구현
- [x] 커버리지 100% 달성
- [x] About 페이지 구현
- [x] Not found 페이지 구현
- [x] 레스토랑 상세페이지 구현

### API
- 레스토랑 지역 목록 보기: [https://eatgo-customer-api.ahastudio.com/regions](https://eatgo-customer-api.ahastudio.com/regions)
- 레스토랑 카테고리 목록 보기: [https://eatgo-customer-api.ahastudio.com/categories](https://eatgo-customer-api.ahastudio.com/categories)
- 레스토랑 목록 보기: [https://eatgo-customer-api.ahastudio.com/restaurants?region=${regionName}&category=${categoryId}](https://eatgo-customer-api.ahastudio.com/restaurants?region=${regionName}&category=${categoryId})
- 레스토랑 상세 정보 보기: [https://eatgo-customer-api.ahastudio.com/restaurants/${restaurantId}](https://eatgo-customer-api.ahastudio.com/restaurants/$%7BrestaurantId%7D)

TaeWoongKim2 committed:
git commit test

TaeWoongKim2 committed:
[Router] 강의를 듣고 라우터 적용

코드숨 강의를 듣고 라우터에 대한 기본적인 사용법 습득.
*TODO: git reverit 후, react-router-dom을 TDD로 다시 구현!

TaeWoongKim2 committed:
Revert "[Router] 강의를 듣고 라우터 적용"

This reverts commit 0e1aa0c0afb7cbd9c3367129f8bc8e9e2fd26b22.

TaeWoongKim2 committed:
[SETTING] webpack.config.js

devServer: { historyApiFallback : { index: 'index.html'  } } 설정값 추가 완료.

TaeWoongKim2 committed:
[Refactor] RestaurantsPage 컴포넌트 생성

Refactor App component.
*TODO: Make test case about RestaurantsPage!

TaeWoongKim2 committed:
[Red] RestaurantsPage에 대한 테스트 생성

1st, 기존 App에 대한 테스트 copy/paste.
2nd, App에 대한 테스트 간단하게 변경.

TaeWoongKim2 committed:
[Red] HomePage에 대한 테스트 작성

Application의 HomePage 컴포넌트에 대한 테스트 작성.
*TODO: Make HomPage component!

TaeWoongKim2 committed:
[Green] HomePage 구현

Make HomePage component.
*TODO: Make tests about AboutPage!

TaeWoongKim2 committed:
[Red] AboutPage에 대한 테스트 작성

AboutPage 컴포넌트에 대한 기본적인 테스트 작성.
*TODO: Make AboutPage component!

TaeWoongKim2 committed:
[Green] AboutPage 구현

테스트를 기반으로 About component 구현.
*TODO: Make NotFoundPage component!

TaeWoongKim2 committed:
[Red] NotFoundPage 테스트 작성

NotFoundPage에 대한 테스트 작성
*TODO: Make NotFoundPage component!

TaeWoongKim2 committed:
[Green] NotFoundPage 컴포넌트 구현

테스트를 기반으로 NotFoundPage 구현
*TODO: Refactor App and Apply React Router

TaeWoongKim2 committed:
[Refactor] Apply react router

1st, intall 'react-router-dom'.
2nd, App component refactor for router.
*TODO: SPA - used to Link in 'react-router-dom'!

TaeWoongKim2 committed:
[Refactor] Apply Link like SPA

<Link to=""> in HomePage component.
*TODO: [Red] HomePage에 대한 테스트 수정!

TaeWoongKim2 committed:
[Red] HomePage에 대한 신규 테스트 추가

1st, MemoryRouter 추가.
2nd, About 및 Restaurants 텍스트 존재유무 테스트.

*TODO: Refector router in App component to index.jsx!

TaeWoongKim2 committed:
[Red] App에 대한 테스트 추가

MemoryRouter 추가.

TaeWoongKim2 committed:
[Red] App router에 대한 테스트 추가

1st, with path '/' case.
2nd, with path '/about' case.
3rd, with path '/restaurants' case.
4th, with undefined path case.

*TODO: Make Header for link to home.

TaeWoongKim2 committed:
[Refactor] Header 영역 구현

Make Header area.
*TODO: Assn.1 to proceed!

TaeWoongKim2 commented:
## 한 일

- 코드숨 6강 라우팅(Routing) 강의를 듣고 라우터에 대한 기본 사용법 익히기!
- 원복 후 처음부터 다시 진행해보기!

## 할 일

- [ ] 레스토랑 상세 페이지 구현하기
- [ ] TDD로 기능 구현
- [ ] 커버리지 100% 달성

Kihyun92 reviewed (commented):
열심히 하려는 의지가 느껴지는 것 같아요 좋습니다~ 👍

`Red-Green-Refactoring`을 잘 적용해보고 계신 것 같네요~ 남겨주신 커밋을 보니 `Refactoring`과정이 빠져있거나, 순서가 맞지 않는 경우도 보여요 ㅎㅎ 테스트의 사이클을 조금 작게 가져가는 것도 좋을 것 같습니다. 원인은 `Green`단계에서 완성도를 높이는데 집중하기 때문일 수도 있을 것 같네요. 당장은 테스트를 성공시키는데 목적을 두고, 그 이후의 완성도는 refactoring 단계에서 챙기는걸 의식하면서 해보셔도 좋을 것 같습니다~ 이래저래 말이 길어졌네요 ㅎㅎ 잘못됐다는 말은 아니니 오해는 말아주세요 😁

TaeWoongKim2 committed:
[Red] 레스토랑 상세정보 API 테스트 추가

API 레스토랑 상세정보에 대한 API 테스트 추가.
*TODO: Make 'fetchRestaurant' method!

TaeWoongKim2 committed:
[Refactor] fetchRestaurant에 대한 테스트 리팩토링.

1st, 레스토랑 상세정보 API의 임시 결과 작성(fixture).
2nd, 테스트 분리.

*TODO: RestaurantPage 라우팅!

TaeWoongKim2 committed:
[Red] App route에 대한 신규 테스트 추가

1st, 'with path /restaurants/:restaurantId'에 대한 테스트 추가.

*TODO: App 라우터에 RestaurantPage 페이지 등록!

TaeWoongKim2 committed:
[Green] RestaurantPage에 대한 라우터 등록

RestaurantPage에 대한 라우터를 등록했으나, 빈페이지 출력됨.

*TODO: RestaurantsContainer에 Link 달기!

TaeWoongKim2 committed:
[Refactor] RestaurantsContainer에 Link 추가로 인한 리팩토링

1st, <Link to=""></Link> 추가.
2nd, RestaurantsContainer 테스트에 MemoryRouter 추가.
3rd, RestaurantsPage 테스트에 MemoryRouter 추가.

*TODO: Link 정상동작 테스트 및 RestaurantPage 테스 작성 및 구현!

TaeWoongKim2 committed:
[Green] RestaurantPage 컴포넌트 구현

1st, Make RestaurantPage component.
2nd, Change Param name restaurantId -> id.
3nd, Refactor Route option 'exact' into App.
4th, Refactor RestarantPage param name 'id'.

*TODO: RestaurantContainer 테스트 및 구현!

TaeWoongKim2 committed:
[Refactor] action creator 'loadRestaurantInfomation' 생성

레스토랑 상세정보를 가져오기 위한 액션 및 디스패치 생성.

*TODO: action.js에 loadRestaurantInfomation 생성!

TaeWoongKim2 committed:
[Red] loadRestaurant에 대한 신규 테스트 추가

1st, mockStore(RESTAURANT);
2nd, await store.dispatch(loadRestaurant());
3rd, runs setRestaurant dispatch.

*TODO: action.js 구현: loadRestaurant & setRestaurant!

TaeWoongKim2 committed:
[Green] loadRestaurant & setRestaurant 에 대한 action 구현

1st, loadRestaurant & setRestaurant 구현.
2nd, api mock method 생성.

*TODO: setRestaurant에 대한 테스트 작성 및 구현!

TaeWoongKim2 committed:
[Red] reducer에 setRestaurant에 대한 테스트 추가

1st, changes restaurant 에 대한 테스트 추가.
2nd, 오류 발생.

*TODO: reducer에 setRestaurant 구현!

TaeWoongKim2 committed:
[Green] reducer에 setRestaurant를 신규 구현

*TODO: refactor reducer.test.jsx!

TaeWoongKim2 committed:
[Refactor] reducer에 대한 테스트 리팩토링

1st, initialState에 restaurant: {}, 추가.
2nd, 오타 수정.

*TODO: RestaurantPage 정상작동 테스트 및 자식 컴포넌트에 대한 테스트 작성!

TaeWoongKim2 committed:
[Refactor] RestaurantPage 리팩토링

loadRestaurant 함수에 id 파라미터 추가.
이에 따라, action, action.test, api 전부 리팩토링.

*TODO: RestaurantPage의 자식 컴포넌트 테스트 작성!

TaeWoongKim2 committed:
[Red] RestaurantContainer에 대한 테스트 작성

레스토랑 상세정보의 컨테이너인 RestaurantContainer 테스트 작성.

*TODO: Make RestaurantContainer component!

TaeWoongKim2 committed:
[Refactor] RestaurantContainer에 redux 상태값 가져옴

useSelector를 통해 restaurant 상태값 가져옴.

TaeWoongKim2 committed:
[Green] RestaurantContainer에 메뉴리스트 추가 구현

1st, menu items에 따른 메뉴 목록 추가 구현.
2st, restaurant에 대한 상세정보가 없을 시, 오류 메시지 출력.

*TODO:
1st, Apply when received code review, and refactoring my code!
2nd, Coverage 100%

Kihyun92 commented on src/services/api.js:
```diff
@@ -19,3 +19,11 @@ export async function fetchRestaurants({ regionName, categoryId }) {
 }
+
+export async function fetchRestaurant({ id }) {
+  const url = 'https://eatgo-customer-api.ahastudio.com/restaurants'
+    + `/${id}`;
```
전체 url에 template literal을 적용하는건 어떨까요?
```suggestion
  const url = `https://eatgo-customer-api.ahastudio.com/restaurants/${id}`;
```

Kihyun92 reviewed (changes requested):
TDD 사이클을 잘 지키면서 구현하신 것 같아요 ㅎㅎ
jest에서 내뿜는 에러들이 많다면 한번에 너무 많은 기능을 테스트 하려한게 아닐까요? 테스트를 조금 더 적게 가져가는 방법이 있을 것 같아요 ㅎㅎ 물론 에러 자체가 굉장히 길다보니 힘들기도 하죠😂 그럴때일 수록 차근차근 해결해보시면 좋을 것 같습니다! 
그리고 TDD 사이클이 헝클어지는건 문제가 아닌것 같아요. 다만 `Red-Green-Refactoring`사이클을 가져가지 못하는 원인에 대해서 고민해보시고, 다음 테스트는 더 잘 구현하면 될 것 같아요 :)

추가로 `RestaurantContainer`를 작업하면서 많은 고민을 해보신 것 같아 좋습니다 👍 
지금도 좋지만, 제 생각에는 이전에 배웠던 `Container/Prensentation component` 측면에서 봤을때, 부모 컴포넌트와 함께 봤을때 `RestaurantContainer` 컴포넌트를 분리 할 수 있을 것 같은데 어떤가요?☺️

TaeWoongKim2 committed:
[Refactor] URL에 template literal 적용

가독성을 위해 template literal 적용.
*TODO: Coverage 100% 달성!

TaeWoongKim2 committed:
[Red] RestaurantContainer에 대한 신규 테스트 추가

1st, 레스토랑 상세설명 렌더링 부분에 대한 테스트 리팩토링.
2nd, 레스토랑에 대한 상세설명이 없을 시, 메시지 출력에 대한 테스트 추가.

*TODO: 코드 리팩토링!

TaeWoongKim2 committed:
[Refactor] 헤더 텍스트 변경

테스트 통과를 위한 헤더 텍스트 변경.
Home -> 헤더.

*TODO: Page Not found 텍스트 변경!

TaeWoongKim2 committed:
[Refactor] Page Not Found 텍스트 변경

테스트 통과를 위한 404 페이지 텍스트 변경.
'404 - Page Not Found' -> '404 Not Found'.

TaeWoongKim2 committed:
[Red] HomePage에 Home 제목 테스트 추가

TaeWoongKim2 committed:
[Green] HomePage에 제목 추가

'Home' 제목 텍스트 추가 (테스트 통과를 위한)

TaeWoongKim2 committed:
[Red] 누락된 RestaurantPage에 대한 테스트 추가

TaeWoongKim2 committed:
[Red] Restaurant에 대한 테스트 추가

RestaurantContainer에서 Restaurant 컴포넌트 분리.

TaeWoongKim2 committed:
[Green] Restaurant 컴포넌트 구현

Restaurant component 구현 및 RestaurantContainer에 Restaurant 분리.

TaeWoongKim2 committed:
[Refactor] Make Loading Message

레스토랑 상세정보 페이지에 대한 로딩 메시지 생성.

TaeWoongKim2 committed:
[Refector] RestaurantContainer의 isEmpty 리팩토링

Function named 'isEmpty' changes to a anonymous function.

TaeWoongKim2 commented:
## 📦 한 일

- 레스토랑 상세정보 페이지,  Container 및 Prensentation 컴포넌트로 분리
- 레스토랑 상세정보 로드 시, 로딩 메시지 출력.
- TDD 방식 구현
- 커버리지 100% 달성

## ⌨️ 할 일

- [ ] 레스토랑 상세정보 URL 직접 접근 및 refresh 시, 출력 안되는 현상 파악


> 모든 테스트 및 기능을 완료 후, ConceptJS 테스트에서 버그를 발견했습니다.
해당 문제는 레스토랑 상세정보 URL(/restaurants/:id)로 직접 접근하거나 해당 페이지에서 새로고침 시, 페이지가 먹통이 되는 현상이 발생했습니다. 해당 문제가 왜 발생하는지 잘 모르겠네요...ㅠㅠ

TaeWoongKim2 committed:
[Refactor] 새로고침 버그 문제 해결

webpack.config.js 에 'html-webpack-plugin' 플러그인 추가.
1st, output: { publicPath: '/' } 추가.
2nd, plugins: [new HtmlWebpackPlugin({ template: './index.html' })] 추가.

TaeWoongKim2 commented:
## 📦 한 일

- 레스토랑 상세정보 URL 직접 접근 및 refresh 시, 출력 안되는 현상 해결

URL(예:/about/test or /restaurants/1)  직접 입력 혹은 새로고침 시, main.js 에 대한 위치를 제대로 불러오지 못하는 것이 원인!

이를 해결하기 위해 '`html-webpack-plugin`' 플러그인을 설치 및 적용.
[참고] https://webpack.js.org/guides/output-management/

이후, 문제가 해결됐습니다.

Kihyun92 reviewed (approved):
이번 주 과제도 쉽지 않았는데 잘해주신것 같습니다 :) TDD도 잘 가져가신 것 같고, 이전 과제에서 배웠던 것들도 잘 적용해주신것 같아요 👍  그리고 커밋 메세지를 잘 작성해주셔서 저도 수월하게 리뷰 드릴 수 있었던 것 같습니다 ㅎㅎ

이번 과제를 하면서 느낀점이나 아쉬웠던 부분, 더 해보고 싶은 부분이 있나요? 한번 고민해보시면 좋겠습니다~

TaeWoongKim2 commented:
1. **한 일(Facts)**
    - **라우팅 처리 방식 및 History API 에 대한 이해**
    리액트 라우터(react-router-dom)를 사용하는 법과 공식 문서를 리딩했다.
    History API 가 어떻게 사용되는지 정리했다!
    - **과제 6. 레스토랑 상세 페이지 구현**
    라우팅을 통해 레스토랑의 상세정보를 출력하도록 구현했다!
    TDD 방식으로 구현했다!
    - **TDD Cycle 반복 연습 (반복숙달중)**

2. **배운 점 (Findings)**
    - **리액트 라우터 처리**
SPA 서비스를 만듬에 있어서 라우터의 역할이 중요하다. 페이지 이동시, 비동기실으로 변경이 필요한 컴포넌트만 재조립하는 방식을 취한다. 동기식에 비해 불필요한 리소스를 다시 로드할 필요가 없기 때문에 불필요한 자원 낭비를 막고, 사용자 경험(UX)도 긍정적으로 높일 수 있는 방식이다.
    기존 vue-router를 이미 사용해봤고 거의 비슷했기 때문에 비교적 수월하게 공식 문서를 읽을 수 있었다.
    - **Script, Style 등 리소스 파일에 대한 경로 이슈**
라우터 처리 시 주의할 점. Depth가 2 이상인 경로 또는 그러한 경로에서 새로고침 시, 스크립트 리소스에 대해 404 에러가 떨어짐을 확인했다. 이는 리소스 파일에 대한 경로가 잡혀있지 않아서 발생했다. 이를 'html-webpack-plugin' 를 통해 해결했다.


> 지금은 페이지가 별로 없고 페이지에서 페이지로 들어가는 깊이가 상당히 얕다고 생각합니다.
페이지가 많을수록 그리고 깊이가 깊을수록 더 가독성 좋고 효율적인 라우터를 구현할 수 있을지 고민하게 되네요!
